src/plasma/plasma_store.c

```diff
@@ -208,42 +234,232 @@ int create_object(client *client_context,
-    result->data_offset = entry->offset;
-    result->metadata_offset = entry->offset + entry->info.data_size;
-    result->data_size = entry->info.data_size;
-    result->metadata_size = entry->info.metadata_size;
-    /* If necessary, record that this client is using this object. In the case
-     * where entry == NULL, this will be called from seal_object. */
-    add_client_to_object_clients(entry, client_context);
-    return OBJECT_FOUND;
-  } else {
-    object_notify_entry *notify_entry;
-    LOG_DEBUG("object not in hash table of sealed objects");
-    HASH_FIND(handle, plasma_state->objects_notify, &object_id,
-              sizeof(object_id), notify_entry);
-    if (!notify_entry) {
-      notify_entry = malloc(sizeof(object_notify_entry));
-      memset(notify_entry, 0, sizeof(object_notify_entry));
-      utarray_new(notify_entry->waiting_clients, &client_icd);
-      notify_entry->object_id = object_id;
-      HASH_ADD(handle, plasma_state->objects_notify, object_id,
-               sizeof(object_id), notify_entry);
+
+  /* Create a get request for this object. */
+  get_request *get_req = malloc(sizeof(get_request));
+  memset(get_req, 0, sizeof(get_request));
+  get_req->client = client_context;
+  get_req->timer = -1;
+  get_req->num_object_ids = num_object_ids;
+  get_req->object_ids = malloc(num_object_ids * sizeof(object_id));
+  get_req->objects = malloc(num_object_ids * sizeof(plasma_object));
+  for (int i = 0; i < num_object_ids; ++i) {
+    get_req->object_ids[i] = object_ids[i];
+  }
+  get_req->num_objects_to_wait_for = num_object_ids;
+  get_req->num_satisfied = 0;
+
+  for (int i = 0; i < num_object_ids; ++i) {
+    object_id obj_id = object_ids[i];
+
+    /* Check if this object is already present locally. If so, record that the
+     * object is being used and mark it as accounted for. */
+    object_table_entry *entry;
+    HASH_FIND(handle, plasma_state->plasma_store_info->objects, &obj_id,
+              sizeof(obj_id), entry);
+    if (entry && entry->state == PLASMA_SEALED) {
+      /* Update the get request to take into account the present object. */
+      initialize_plasma_object(&get_req->objects[i], entry);
+      get_req->num_satisfied += 1;
+      /* If necessary, record that this client is using this object. In the case
+       * where entry == NULL, this will be called from seal_object. */
+      add_client_to_object_clients(entry, client_context);
+    } else {
+      /* Add a placeholder plasma object to the get request to indicate that the
+       * object is not present. This will be parsed by the client. */
+      get_req->objects[i].data_size = -1;
+      /* Add the get request to the relevant data structures. */
+      add_get_request_for_object(plasma_state, obj_id, get_req);
     }
-    utarray_push_back(notify_entry->waiting_clients, &client_context);
   }
-  return OBJECT_NOT_FOUND;
+
+  /* If all of the objects are present already, return to the client. */
```

Comment:
Might be worth it to also return immediately if `timeout_ms` is 0, since this will be a common pattern.